### PR TITLE
Show kiwi image packages list

### PR DESCRIPTION
**Pull request by Ana06**
Kiwi package list is now render in the kiwi::Image show page. :bowtie: For that we:

- Added migrations for Kiwi Packages and introduced the needed validations in the models.
- Added Kiwi Image breadcrumb
- Imported also packages in the `Kiwi::Image#import_from_package` method.
- Add packages list for the package groups with `image` type and `onlyRequired` pattern type to the Kiwi::Image show page. (As it is currently done like that in Studio.

The view looks like:

image

The Edit button would be moved down when editing is also available for packages. Feedback about how to improve the view and the information that should render is welcome (although some of the change may be better done in a different PR to avoid making this too big).

Mob-programmed by @DavidKang and @Ana06, and part with @mdeniz too.

### CODE REVIEW

File: src/api/app/models/kiwi/package_group.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
+class Kiwi::PackageGroup < ApplicationRecord
+  has_many :packages
+  belongs_to :image
+
+  # we need to add a prefix, to avoid generating class methods that already
+  # exist in Active Record, such as "delete"
+  enum kind: %i[bootstrap delete docker image iso lxc oem pxe split testsuite vmx], _prefix: :kind
```
Comment (evanrolfe) [inline src/api/app/models/kiwi/package_group.rb]: [Nitpick] Does the term "kind" come from the kiwi package description?

File: src/api/app/models/kiwi/image.rb [hunk 2]
```diff
@@ -103,6 +125,10 @@ def outdated?
 
     package.kiwi_image_outdated?
   end
+
+  def packages
+    package_groups.where(kind: 'image', pattern_type: 'onlyRequired').map(&:packages).flatten
```
Comment (evanrolfe) [inline src/api/app/models/kiwi/image.rb]: Isn't this a has_many through association?

File: src/api/app/models/kiwi/image.rb [hunk 3]
```diff
@@ -103,6 +125,10 @@ def outdated?
 
     package.kiwi_image_outdated?
   end
+
+  def packages
```
Comment (evanrolfe) [inline src/api/app/models/kiwi/image.rb]: I think we need to distinguish between packages and kiwi_packages. Because at the moment if you do `kiwi_image.packge` you get a `::Package` but if you do `kiwi_image.packages` you get an array of `Kiwi::Package` objects. So I would rename this method to something else maybe `kiwi_packages`?

File: src/api/app/models/kiwi/package_group.rb (hunk 1 above)
Comment (DavidKang) [inline src/api/app/models/kiwi/package_group.rb]: @evanrolfe, we are using `kind` because in the kiwi description we have `type`, so we are using it instead

File: src/api/app/views/webui/kiwi/images/show.html.haml [hunk 4]
```diff
@@ -10,3 +10,18 @@
...
     = button_to('Edit', edit_kiwi_image_path(@image), method: :get)
+
+%h3 Packages
+- if @image.packages.empty?
+  %p There are no packages.
+- else
+  %table.kiwi_packages_table
+    %tr
+      %th Package
+      %th arch
```
Comment (ChrisBr) [inline src/api/app/views/webui/kiwi/images/show.html.haml]: I would completely remove these columns and only show for now the package name and maybe a link to details. As you already see from your screenshot, it is very rare that you have additional information / attributes (bootinclude is usually only 2-3 packages, arch is usually the same for all packages etc etc).

File: src/api/app/models/kiwi/image.rb [hunk 5]
```diff
@@ -62,6 +63,27 @@ def self.build_from_xml(xml_string, md5)
       new_image.repositories.build(attributes)
       order += 1
     end
+    package_groups = xml["packages"]
+    package_groups = [xml["packages"]] if xml["packages"].is_a?(Hash)
+    package_groups.each do |package_group_xml|
+      attributes = {
+        kind:         package_group_xml['type'],
```
Comment (evanrolfe) [inline src/api/app/models/kiwi/image.rb]: Why not just use type for the attribute name?

File: src/api/app/models/kiwi/package_group.rb (hunk 1 above)
Comment (evanrolfe) [inline src/api/app/models/kiwi/package_group.rb]: But if its called `type` in the kiwi description why not just keep that naming and call it `type` in the model too?

File: src/api/Rakefile [hunk 6]
```diff
@@ -29,7 +29,7 @@ unless Rails.env.production?
   require 'haml_lint/rake_task'
 
   RuboCop::RakeTask.new(:rubocop) do |task|
-    task.options = ['-D', '-F', '--fail-level', 'convention']
+    task.options = ['-D', '-a']
```
Comment (bgeuken) [inline src/api/Rakefile]: This looks unrelated. If this was an intended change, please move it to another PR.

Comment (Ana06) [inline src/api/Rakefile]: Was not intended, thanks!

File: src/api/app/models/kiwi/image.rb (hunk 5 above)
Comment (Ana06) [inline src/api/app/models/kiwi/image.rb]: because type is used by rails for Single Table Inheritance

File: src/api/app/models/kiwi/image.rb (hunk 2 above)
Comment (Ana06) [inline src/api/app/models/kiwi/image.rb]: and how can you write the where in the association? :thinking:

File: src/api/app/models/kiwi/package_group.rb (hunk 1 above)
Comment (Ana06) [inline src/api/app/models/kiwi/package_group.rb]: because type is used by rails for Single Table Inheritance

File: src/api/app/models/kiwi/image.rb (hunk 5 above)
Comment (evanrolfe) [inline src/api/app/models/kiwi/image.rb]: You can disable STI by setting this in the model:
```ruby
self.inheritance_column = nil
```
What does everybody think about that? To me it makes more sense to do this but maybe others have different opinions?

File: src/api/app/models/kiwi/image.rb (hunk 2 above)
Comment (evanrolfe) [inline src/api/app/models/kiwi/image.rb]: You can add conditions to an association here is an example from the [docs](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html)
```ruby
has_many :approved_comments, -> { where(approved: true) }, class_name: 'Comment'
```

But I wonder, is packages the right name for this if its only returning packages that are of type image and pattern_type onlyRequired? Maybe the association should be named "required_image_packages"? I dont really know the context of this stuff so I might be wrong..

File: src/api/app/models/kiwi/image.rb [hunk 7]
```diff
@@ -62,6 +63,27 @@ def self.build_from_xml(xml_string, md5)
...
+        pattern_type: package_group_xml['patternType']
+      }
+      package_group = Kiwi::PackageGroup.new(attributes)
+      package_group_xml['package'].each do |package|
+        attributes = {
+          name:     package['name'],
+          arch:     package['arch'],
+          replaces: package['replaces']
+        }
+        attributes['bootinclude'] = package['bootinclude'] == 'true' if package.key?('bootinclude')
```
Comment (ChrisBr) [inline src/api/app/models/kiwi/image.rb]: The ```if package.key?('bootinclude')``` isn't necessary, is it?

```
irb(main):001:0> foo = {}
=> {}
irb(main):002:0> foo['bootinclude']
=> nil
irb(main):003:0> foo['bootinclude'] == 'true'
=> false
irb(main):004:0> 

```

File: src/api/app/models/kiwi/image.rb [hunk 8]
```diff
@@ -62,6 +63,27 @@ def self.build_from_xml(xml_string, md5)
...
+        attributes['bootdelete'] = package['bootdelete'] == 'true' if package.key?('bootdelete')
```
Comment (ChrisBr) [inline src/api/app/models/kiwi/image.rb]: Same here

File: src/api/app/models/kiwi/image.rb (hunk 2 above)
Comment (ChrisBr) [inline src/api/app/models/kiwi/image.rb]: Here is the documentation:
https://doc.opensuse.org/projects/kiwi/doc/

Not sure if it is really necessary to distinguish between the pattern / type etc for now. 
Maybe we can a scope?

File: src/api/app/models/kiwi/package_group.rb (hunk 1 above)
Comment (bgeuken) [inline src/api/app/models/kiwi/package_group.rb]: kiwi_type would still be an option. Though I have no strong opinion about this.

File: src/api/app/models/kiwi/image.rb (hunk 5 above)
Comment (ChrisBr) [inline src/api/app/models/kiwi/image.rb]: Agree with @bgeuken to use kiwi_type to avoid the mess with disable inheritance_column...

File: src/api/app/models/kiwi/image.rb (hunk 7 above)
Comment (Ana06) [inline src/api/app/models/kiwi/image.rb]: to avoid storing as `false` something that is actually `nil`.

**Comment by Ana06** [comment]
Changes made! @evanrolfe @ChrisBr @bgeuken please re-review it!

The UI looks currently like:

image

We plan to make a new PR to improve the UI and to allow edit packages, but we want to avoid to make this PR too big.
